src/rdysetup.c

```diff
@@ -832,15 +832,150 @@ static PetscErrorCode InitSolution(RDy rdy) {
+    Vec                  local       = NULL;
+    if (sediment_ic.file[0]) {  // read sediment data from file
+      PetscViewer viewer;
+      PetscCall(PetscViewerBinaryOpen(rdy->comm, sediment_ic.file, FILE_MODE_READ, &viewer));
+
+      Vec natural, global;
+      PetscCall(DMPlexCreateNaturalVector(rdy->sediment_dm, &natural));
+      PetscCall(DMCreateGlobalVector(rdy->sediment_dm, &global));
+      PetscCall(DMCreateLocalVector(rdy->sediment_dm, &local));
+
+      PetscCall(VecLoad(natural, viewer));
+      PetscCall(PetscViewerDestroy(&viewer));
+
+      // check the block size of the initial condition vector agrees with the block size of rdy->u_global
+      PetscInt nblocks_nat;
+      PetscCall(VecGetBlockSize(natural, &nblocks_nat));
+      PetscCheck((ndof == nblocks_nat), rdy->comm, PETSC_ERR_USER,
+                 "The block size of the initial condition ('%" PetscInt_FMT
+                 "') "
+                 "does not match with the number of DOFs ('%" PetscInt_FMT "')",
+                 nblocks_nat, ndof);
+
+      // scatter natural-to-global
+      PetscCall(DMPlexNaturalToGlobalBegin(rdy->sediment_dm, natural, global));
+      PetscCall(DMPlexNaturalToGlobalEnd(rdy->sediment_dm, natural, global));
+
+      // scatter global-to-local
+      PetscCall(DMGlobalToLocalBegin(rdy->sediment_dm, global, INSERT_VALUES, local));
+      PetscCall(DMGlobalToLocalEnd(rdy->sediment_dm, global, INSERT_VALUES, local));
+
+      // free up memory
+      PetscCall(VecDestroy(&natural));
+      PetscCall(VecDestroy(&global));
+    }
+
+    // set regional sediment as needed
+    for (PetscInt r = 0; r < rdy->num_regions; ++r) {
+      RDyRegion    region = rdy->regions[r];
+      RDyCondition ic     = rdy->initial_conditions[r];
+      if (!strcmp(ic.sediment->name, sediment_ic.name)) {
+        if (local) {
+          PetscScalar *local_ptr;
+          PetscCall(VecGetArray(local, &local_ptr));
+          for (PetscInt c = 0; c < region.num_local_cells; ++c) {
+            PetscInt cell_local_id = region.cell_local_ids[c];
+            PetscInt owned_cell_id = rdy->mesh.cells.local_to_owned[cell_local_id];
+            if (rdy->mesh.cells.is_owned[cell_local_id]) {  // skip ghost cells
+              for (PetscInt idof = 0; idof < ndof; idof++) {
+                u_ptr[ndof * owned_cell_id + idof] = local_ptr[ndof * cell_local_id + idof];
+              }
+            }
+          }
+          PetscCall(VecRestoreArray(local, &local_ptr));
+          PetscCall(VecDestroy(&local));
+        } else {
+          for (PetscInt c = 0; c < region.num_owned_cells; ++c) {
+            PetscInt owned_cell_id      = region.owned_cell_global_ids[c];
+            u_ptr[ndof * owned_cell_id] = mupEval(sediment_ic.concentration);
+          }
+        }
+      }
+    }
+  }
+
+  PetscCall(VecRestoreArray(rdy->sediment_u_global, &u_ptr));
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+static PetscErrorCode InitFlowAndSedimentSolution(RDy rdy) {
```

Comment:
Should we call this `InitTotalSolution`, in anticipation of adding salinity?

Reply:
Ah, never mind. I see now that this is done w.r.t. sediments in `InitSolution` below.